azerothcore/azerothcore-wotlk
# feat(Core/Config): rework configs load system

PR based @BarbzYHOOL idea in PR (#2560)

## CHANGES PROPOSED:
* Add new function for load modules configs
* Add messages for the modules using config variant (see screenshots)
* Rework loading config method
* Replace type in functions `Get***Default` from `const char* name` to `std::string const& name` (for new logging system)
* Now the loading of configs modules is separated from the configs servers.  See `LoadAppConfigs()` and `LoadModulesConfigs();`

## About loading config method
### Before
1. Load `conf..dist` conf
```cpp
if (!sConfigMgr->LoadInitial(cfg_def_file.c_str())) 
{
    printf("ERROR: Invalid or missing default configuration file : %s\n", cfg_def_file.c_str());
    return 1;
}
```
2. Load `.conf` conf
```cpp
if (!sConfigMgr->LoadMore(cfg_file))
{
    printf("WARNING: Invalid or missing configuration file : %s\n", cfg_file);
    printf("Verify that the file exists and has \'[worldserver]' written in the top of the file!\n");
}
```
3. Add list configs in `World` class in `game` library
```cpp
//set module config file list
sWorld->SetConfigFileList(CONFIG_FILE_LIST);
```
4. Load modules configs in fuction `SetInitialWorldSettings`
```cpp
///- Initialize the World
sWorld->SetInitialWorldSettings();
```
### After
1. Add list configs files (server and modules if app == worldserver)
```cpp
sConfigMgr->SetConfigList(std::string(configFile), std::string(CONFIG_FILE_LIST));
```
2. Load apps configs
```cpp
if (!sConfigMgr->LoadAppConfigs())
   return 1;
```
3. Load modules config (in Config class) after loaded db
```cpp
sConfigMgr->LoadModulesConfigs();
```
---

## TESTS PERFORMED:
- Build and run in Win at VS `16.7.0 Preview 3.1`
- Build and run in Linux `Debian10` at `Clang7`

## HOW TO TEST THE CHANGES:
- Build and run auth/worldserver with configs modules
- Check logs in console/file
- See screenshots for example

## Target branch(es):

## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### Media
- Screenshots
![image](https://user-images.githubusercontent.com/18329778/72204792-64e67480-34ae-11ea-8c92-60c6d2aa906a.png)
![image](https://user-images.githubusercontent.com/18329778/72204761-fd302980-34ad-11ea-8834-2863d67b19b1.png)


```
2020-01-11 20:06:54 Using configuration file worldserver.conf.
2020-01-11 20:06:54 Using SSL version: OpenSSL 1.1.1d  10 Sep 2019 (library: OpenSSL 1.1.1d  10 Sep 2019)
2020-01-11 20:06:54 Using ACE version: 6.3.2
2020-01-11 20:06:54 AzerothCore rev. 97c02b70f922+ 2020-01-11 19:40:31 +0700 (ac-rework-config branch) (Win64, Release) (worldserver-daemon)
2020-01-11 20:06:54 <Ctrl-C> to stop.

2020-01-11 20:06:54    █████╗ ███████╗███████╗██████╗  ██████╗ ████████╗██╗  ██╗
2020-01-11 20:06:54   ██╔══██╗╚══███╔╝██╔════╝██╔══██╗██╔═══██╗╚══██╔══╝██║  ██║
2020-01-11 20:06:54   ███████║  ███╔╝ █████╗  ██████╔╝██║   ██║   ██║   ███████║
2020-01-11 20:06:54   ██╔══██║ ███╔╝  ██╔══╝  ██╔══██╗██║   ██║   ██║   ██╔══██║
2020-01-11 20:06:54   ██║  ██║███████╗███████╗██║  ██║╚██████╔╝   ██║   ██║  ██║
2020-01-11 20:06:54   ╚═╝  ╚═╝╚══════╝╚══════╝╚═╝  ╚═╝ ╚═════╝    ╚═╝   ╚═╝  ╚═╝
2020-01-11 20:06:54                                 ██████╗ ██████╗ ██████╗ ███████╗
2020-01-11 20:06:54                                 ██╔════╝██╔═══██╗██╔══██╗██╔═══╝
2020-01-11 20:06:54                                 ██║     ██║   ██║██████╔╝█████╗
2020-01-11 20:06:54                                 ██║     ██║   ██║██╔══██╗██╔══╝
2020-01-11 20:06:54                                 ╚██████╗╚██████╔╝██║  ██║███████╗
2020-01-11 20:06:54                                  ╚═════╝ ╚═════╝ ╚═╝  ╚═╝╚══════╝

2020-01-11 20:06:54   	  AzerothCore 3.3.5a  -  www.azerothcore.org

2020-01-11 20:06:54 Realm running as realm ID 1
2020-01-11 20:06:54 Using World DB: ACDB 335.3 (dev)
2020-01-11 20:06:54 
2020-01-11 20:06:54 Load config error. Invalid or missing configuration file: worldserver.conf
2020-01-11 20:06:54 Verify that the file exists and has '[worldserver]' written in the top of the file!
2020-01-11 20:06:54 
2020-01-11 20:06:54 ERROR: Load config error. Invalid or missing dist configuration file: ExternalMail.conf.dist
2020-01-11 20:06:54 ERROR: Verify that the file exists and has '[worldserver]' written in the top of the file!
2020-01-11 20:06:54 
2020-01-11 20:06:54 Load config error. Invalid or missing configuration file: worldserver.conf
2020-01-11 20:06:54 Verify that the file exists and has '[worldserver]' written in the top of the file!
2020-01-11 20:06:54 
2020-01-11 20:06:54 Load config error. Invalid or missing configuration file: worldserver.conf
2020-01-11 20:06:54 Verify that the file exists and has '[worldserver]' written in the top of the file!
2020-01-11 20:06:54 
2020-01-11 20:06:54 Load config error. Invalid or missing configuration file: worldserver.conf
2020-01-11 20:06:54 Verify that the file exists and has '[worldserver]' written in the top of the file!
2020-01-11 20:06:54 
2020-01-11 20:06:54 Using configuration for modules:
2020-01-11 20:06:54 > Module (transmog) using (transmog.conf)
2020-01-11 20:06:54 > Module (DungeonBuff) using (DungeonBuff.conf.dist)
2020-01-11 20:06:54 > Module (QuestBuff) using (QuestBuff.conf.dist)
2020-01-11 20:06:54 > Module (ExternalMail) using (default hardcoded settings)
2020-01-11 20:06:54 > Module (CFBG) using (CFBG.conf.dist)
2020-01-11 20:06:54 
2020-01-11 20:06:54 Initializing Scripts...
2020-01-11 20:06:54 Loading C++ scripts
2020-01-11 20:06:54 Will clear `logs` table of entries older than 1209600 seconds every 10 minutes.
2020-01-11 20:06:54 Using DataDir ../data/
2020-01-11 20:06:54 WORLD: VMap support included. LineOfSight:1, getHeight:1, indoorCheck:1 PetLOS:1
2020-01-11 20:06:54 ERROR: -> Not found option 'External.Mail.Enable'. The default value is used (No)
2020-01-11 20:06:54 ERROR: -> Not found option 'External.Mail.Interval'. The default value is used (1)
2020-01-11 20:06:54 
2020-01-11 20:06:54 Loading acore strings...
2020-01-11 20:06:54 >> Loaded 1171 acore strings in 4 ms
...
```
